### PR TITLE
Flink: move hadoop configuration to Loaders from Source/Sink API

### PR DESCRIPTION
**Pull request by JingsongLi**
Discussed in: https://github.com/apache/iceberg/pull/1509/files#r496268020

### CODE REVIEW

**Comment by JingsongLi** [comment]
CC: @rdblue @openinx

File: flink/src/main/java/org/apache/iceberg/flink/CatalogLoader.java
Comment (openinx) [inline flink/src/main/java/org/apache/iceberg/flink/CatalogLoader.java:47]: Nice catch, since the `CatalogLoader` need to be serializable, so all members need to be serializable. Do we need to add several unit tests to address the serializable issue so that we won't break it in future ?   similar to the unit test TestScanTaskSerialization ?

Comment (openinx) [reply flink/src/main/java/org/apache/iceberg/flink/CatalogLoader.java:47]: `TableLoader`  need the similar unit test.

File: flink/src/main/java/org/apache/iceberg/flink/TableLoader.java
Comment (openinx) [inline flink/src/main/java/org/apache/iceberg/flink/TableLoader.java:48]: Is this only visible for testing ?  Moving to test package ?

**Review by openinx — COMMENTED** [review]
Skimmed the patch, moving the `Configuration`  to the concrete CatalogLoader or TableLoader really simplify the codes,  that looks a good improvement.

File: flink/src/main/java/org/apache/iceberg/flink/CatalogLoader.java
Comment (JingsongLi) [reply flink/src/main/java/org/apache/iceberg/flink/CatalogLoader.java:47]: This is already tested by ITCases. It is the only way which must be passed.
But unit test can be added too.

File: flink/src/main/java/org/apache/iceberg/flink/TableLoader.java
Comment (JingsongLi) [reply flink/src/main/java/org/apache/iceberg/flink/TableLoader.java:48]: I think it is useful for users, since Flink has default hadoop configuration.

File: flink/src/main/java/org/apache/iceberg/flink/CatalogLoader.java
Comment (chenjunjiedada) [reply flink/src/main/java/org/apache/iceberg/flink/CatalogLoader.java:47]: What do you mean ITCases?

Comment (JingsongLi) [reply flink/src/main/java/org/apache/iceberg/flink/CatalogLoader.java:47]: `TestFlinkScanSql`, `TestFlinkTableSink` and `TestFlinkIcebergSink` will cover this too.

File: flink/src/main/java/org/apache/iceberg/flink/FlinkCatalogFactory.java
Comment (rdblue) [inline flink/src/main/java/org/apache/iceberg/flink/FlinkCatalogFactory.java:70]: It's a little odd that the arguments here are the same as `createCatalog`, but that they are in a different order and use "options" rather than "properties". Should we at least use the same argument order?

File: flink/src/test/java/org/apache/iceberg/flink/TestTableLoader.java
Comment (rdblue) [inline flink/src/test/java/org/apache/iceberg/flink/TestTableLoader.java:32]: I don't think it is sufficient to test that the loader's `toString` representation is the same. This would still pass if the underlying `Configuration` were just set to null and not serialized at all. I think it would make sense to do something that will fail if the `Configuration` is not serialized and deserialized correctly, like load the table. In that case, you would just need to create a Hadoop table on the local FS and load it through the copied loader.

File: flink/src/test/java/org/apache/iceberg/flink/TestCatalogLoader.java
Comment (rdblue) [inline flink/src/test/java/org/apache/iceberg/flink/TestCatalogLoader.java:34]: Same here. I think these tests should do a little more than validate the `toString` method.

File: flink/src/main/java/org/apache/iceberg/flink/CatalogLoader.java
Comment (rdblue) [reply flink/src/main/java/org/apache/iceberg/flink/CatalogLoader.java:47]: I'm not an expert on how Flink will run in tests, but I think it would be safer not to rely on end-to-end or integration tests to check serialization because there may not be a guarantee that tasks are serialized for those runners. (And this was a problem for Hive testing in the past.)

**Comment by rdblue** [comment]
@JingsongLi, thanks for working on this! The implementation looks great, but I think it would be good to have a bit more thorough  validation in the tests.

File: flink/src/main/java/org/apache/iceberg/flink/CatalogLoader.java
Comment (JingsongLi) [reply flink/src/main/java/org/apache/iceberg/flink/CatalogLoader.java:47]: Flink `StreamExecutionEnvironment.execute` will launch a local cluster, the client will connect to this cluster by RPC through binary message.
I think you are right, it is better to add unit test too.

File: flink/src/main/java/org/apache/iceberg/flink/FlinkCatalogFactory.java
Comment (JingsongLi) [reply flink/src/main/java/org/apache/iceberg/flink/FlinkCatalogFactory.java:70]: I will use same order and `options`.

File: flink/src/test/java/org/apache/iceberg/flink/TestTableLoader.java
Comment (JingsongLi) [reply flink/src/test/java/org/apache/iceberg/flink/TestTableLoader.java:32]: Agree

File: flink/src/main/java/org/apache/iceberg/flink/IcebergTableSource.java
Comment (openinx) [inline flink/src/main/java/org/apache/iceberg/flink/IcebergTableSource.java:73]: nit:   I see somewhere call it `options` and other place call it `properties`, is it possible to unify them ?

File: flink/src/test/java/org/apache/iceberg/flink/TestCatalogTableLoader.java
Comment (openinx) [inline flink/src/test/java/org/apache/iceberg/flink/TestCatalogTableLoader.java:88]: Missed a test for `HiveTableLoader` ?  I think we don't have to open the table actually,  just need to ensure the members inside the instance are the same, maybe adding an equals inside the CatalogLoader or TableLoader.

File: flink/src/main/java/org/apache/iceberg/flink/IcebergTableSource.java
Comment (JingsongLi) [reply flink/src/main/java/org/apache/iceberg/flink/IcebergTableSource.java]: OK I'll do it.

File: flink/src/test/java/org/apache/iceberg/flink/TestCatalogTableLoader.java
Comment (JingsongLi) [reply flink/src/test/java/org/apache/iceberg/flink/TestCatalogTableLoader.java:88]: You mean `Tableloader.CatalogTableloader` class? I think we don't need to add case for it since we have case for `Catalog Loader`s.
Opening the table can cover more since they are loaders, we can try to load something and ensure that is correct.

**Review by rdblue — APPROVED** [review]

**Comment by rdblue** [comment]
I think this is ready, but @openinx will probably want to have a look at the answer to his question about a test for `HiveTableLoader` so I'll wait for him to merge. Thanks, @JingsongLi!

File: flink/src/test/java/org/apache/iceberg/flink/TestCatalogTableLoader.java
Comment (openinx) [reply flink/src/test/java/org/apache/iceberg/flink/TestCatalogTableLoader.java:88]: OK, it's 'CatalogTableLoader'.  it has two members: `CatalogLoader` & `String identifier` (the other one is transient member). I agree that in this case if we could assert that `CatalogLoader` is serializing & deserializing correctly, then it imply that `CatalogTableLoader` should be OK. But providing those unit tests is to ensure that we won't break this point in future.  What if others introduce another non-serializable member in `CatalogTableLoader` class?  that's why I said it seems to be more reasonable to write a ser/der unit test for `CatalogTableLoader`.

Comment (JingsongLi) [reply flink/src/test/java/org/apache/iceberg/flink/TestCatalogTableLoader.java:88]: make sense, I'll add test.

**Comment by openinx** [comment]
LGTM now, will merge this if all unit tests pass.

**Comment by JingsongLi** [comment]
Thanks @rdblue and @openinx for review, merging...

**Pull request merged**
